Support horizontal HistogramLUT{Item,Widget}

ixjlyons opened the pull request:
A few things going on here (sorry):

- Support for horizontal orientation (fixes #99)
- Documentation improvements (and conversion to numpy-style docstrings)
- A few minor bug fixes and cosmetic-ish code changes

This felt very ugly to implement, so suggestions for cleaning it up would be appreciated.

I also thought perhaps we could just assume the orientation should be vertical if gradient position is left or right and horizontal otherwise. However, from a user perspective, I'd probably rather specify orientation and let the default gradient position take effect rather than use the gradient position to control orientation.

ixjlyons commented on pyqtgraph/graphicsItems/HistogramLUTItem.py:
```diff
@@ -25,177 +26,268 @@
         self.lut = None
         self.imageItem = lambda: None  # fake a dead weakref
         self.levelMode = levelMode
-        self.rgbHistogram = rgbHistogram
```
`git blame` indicates this was added but never implemented

j9ac9k replied:
that would be a cool feature tho!

ixjlyons replied:
I think `levelMode='rgba'` does what this was probably intended to mean. Check `examples/HistogramLUT.py` which has a toggle to switch between mono/rgba.

j9ac9k replied:
oh cool, it is implemented there, that looks really slick

ixjlyons commented on pyqtgraph/graphicsItems/HistogramLUTItem.py:
```diff
@@ -25,177 +26,268 @@
+        mn, mx = self.getLevels()
+        vbc = self.vb.viewRect().center()
         gradRect = self.gradient.mapRectToParent(self.gradient.gradRect.rect())
+        if self.orientation == 'vertical':
```
Not sure if moving this stuff all up here really improves readability compared to a block of if statements in the draw loop. This is a particular spot where it felt like there's got to be a better way

j9ac9k commented on pyqtgraph/graphicsItems/HistogramLUTItem.py:
```diff
@@ -25,177 +26,268 @@
+        self.gradient = GradientEditorItem(orientation=self.gradientPosition)
         self.gradient.loadPreset('grey')
+
+        ori = 'horizontal' if self.orientation == 'vertical' else 'vertical'
```
is this right?  shouldn't it be

```python
ori = 'horizontal if self.orientation == 'horizontal' else 'vertical'
```

ixjlyons replied:
Yeah I considered adding a comment here because it's confusing. For `LinearRegionItem`, orientation refers to the orientation of the *lines*.

j9ac9k replied:
Oh right, may I suggest renaming the variable to `regionOrientation` or something along those lines.

j9ac9k commented on pyqtgraph/graphicsItems/HistogramLUTItem.py:
```diff
@@ -343,15 +444,15 @@ def _showRegions(self):
         else:
-            raise ValueError("Unknown level mode %r" %  self.levelMode) 
-    
+            raise ValueError("Unknown level mode {}".format(self.levelMode))
```
this is an improvement, but since we're on python 3.7 as a minimum, you could use a f-string if you wanted to

ixjlyons replied:
Heh, forgot I even did this. f-strings are good for this kind of thing so I'll go ahead and update this and a couple others.

ixjlyons commented:
Somehow forgot to add the new keyword arg to the docstring. Added.

j9ac9k commented on pyqtgraph/graphicsItems/HistogramLUTItem.py:
```diff
@@ -277,56 +371,66 @@ def getLevels(self):
-                self.regions[i+1].setRegion(levels)
-        
+                rgba = 4*[(min, max)]
+            for levels, region in zip(rgba, self.regions[1:]):
```
just wanted to say I love this change 👍🏻

ixjlyons commented:
> Somehow forgot to add the new keyword arg to the docstring. Added.

I had also forgotten to add 'top', and 'bottom' to the `gradientPosition` doc. Added.

j9ac9k commented:
Thanks @ixjlyons for this PR, this looks good, the improvement in documentation and code-readability is greatly appreciated, merging!

pijyoi commented:
What is the expected behavior when an RGBA image is used?
In the example below, the histograms seem to be "crushed" by the alpha channel.
```python
import numpy as np
import scipy.misc

import pyqtgraph as pg
from pyqtgraph.Qt import QtWidgets

app = pg.mkQApp()
win = QtWidgets.QMainWindow()
win.resize(880, 600)
win.show()

cw = QtWidgets.QWidget()
cw.show()
win.setCentralWidget(cw)

layout = QtWidgets.QGridLayout()
cw.setLayout(layout)
layout.setSpacing(0)

view = pg.GraphicsView()
vb = pg.ViewBox()
vb.setAspectLocked()
vb.invertY(True)
view.setCentralItem(vb)
layout.addWidget(view, 0, 1, 3, 1)

hist = pg.HistogramLUTWidget(gradientPosition="left")
layout.addWidget(hist, 0, 2)

monoRadio = QtWidgets.QRadioButton('mono')
rgbaRadio = QtWidgets.QRadioButton('rgba')
layout.addWidget(monoRadio, 1, 2)
layout.addWidget(rgbaRadio, 2, 2)
monoRadio.setChecked(True)

def setLevelMode():
    mode = 'mono' if monoRadio.isChecked() else 'rgba'
    hist.setLevelMode(mode)

monoRadio.toggled.connect(setLevelMode)

face = scipy.misc.face()
data = np.full_like(face, 255, shape=face.shape[:2] + (4,))
data[..., :3] = face

img = pg.ImageItem(data, axisOrder='row-major')
vb.addItem(img)
vb.autoRange()

hist.setImageItem(img)

app.exec_()
```

ixjlyons commented:
I noticed that as well. The histogram comes from `ImageItem.getHistogram` and, well, it's true that if your alpha channel is all 255s, that bin is going to tower over everything whether the histogram is computed on the flattened image or per channel. Suggestions welcome for how to handle it. I could imagine an option to ignore the alpha channel or allowing more flexibility on the "count" axis of the histogram. Maybe this case is common enough to handle automatically with a special case, i.e. detect when a histogram bin value is >= the image width*height. That wouldn't really be robust for 'mono' mode with an RGB image though. For 'rgba' mode, normalizing the histograms could help if each curve gets its own density axis. Maybe there's something obvious I'm not thinking of.

pijyoi commented:
Well, the question is whether it is meaningful to manipulate the "levels" of the alpha channel, The (common?) use-case for alpha channel seems to be for the whole image to take only 2 distinct values: 0 and some other value (typically 255 or 1.0). One could say that it is wrong usage to supply an RGBA image to HistogramLUT.

ixjlyons commented:
I agree it's probably not too common to want to control or even see the alpha level, but I'm not sure I'd consider it incorrect usage to pair an `ImageItem` with an alpha channel together with a `HistogramLUTItem`. I'm thinking we could add an "rgb" `levelMode`.
